File: 0221/dp.cpp
#include <algorithm>
#include <vector>

#include "leetcode/assert.hpp"

using namespace std;
using namespace LeetCode;

// time complexity: O(mn)
// time complexity: O(n)
class Solution {
 public:
  int maximalSquare(vector<vector<char>>& matrix) {
    const size_t m = matrix.size();
    if (m == 0) { return 0; }
    const size_t n = matrix[0].size();
    vector<int> dp_prev(n+1, 0);
    int ret = 0;
    for (size_t i = 1; i <= m; ++i) {
      vector<int> dp_curr(n+1, 0);
      for (size_t j = 1; j <= n; ++j) {
        if (matrix[i-1][j-1] == '0') { continue; }
        dp_curr[j] = min({dp_curr[j-1], dp_prev[j], dp_prev[j-1]}) + 1;
        ret = max(ret, dp_curr[j]);
      }
      dp_prev.swap(dp_curr);
    }
    return ret * ret;
  }
};

int main() {
  Solution sol;
  vector<vector<char>> matrix;

  matrix = {
    {'1','0','1','0','0'},
    {'1','0','1','1','1'},
    {'1','1','1','1','1'},
    {'1','0','0','1','0'},
  };
  ExpectEqual(4, sol.maximalSquare(matrix));

  matrix = {
    {'0','1','1','1','0'},
    {'1','1','1','0','0'},
    {'1','1','1','1','0'},
    {'1','1','1','1','0'},
  };
  ExpectEqual(9, sol.maximalSquare(matrix));

  matrix = {
    {'0','1','1','1','0'},
    {'1','1','0','0','0'},
    {'0','1','1','0','0'},
    {'1','0','1','1','0'},
  };
  ExpectEqual(1, sol.maximalSquare(matrix));

  matrix = {};
  ExpectEqual(0, sol.maximalSquare(matrix));
}

File: 0264/dp.py
from typing import *


#  i 0 1 2 3 4 5 6
# dp 1 2 3 4 5 6 8
# p2 0 1 1 2 2 3 4
# p3 0 0 1 1 1 2 2
# p5 0 0 0 0 1 1 1
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        dp = [None for _ in range(n)]
        dp[0] = 1
        p2, p3, p5 = 0, 0, 0
        for i in range(1, n):
            dp[i] = min(dp[p2] * 2, dp[p3] * 3, dp[p5] * 5)
            if dp[i] == dp[p2] * 2:
                p2 += 1
            if dp[i] == dp[p3] * 3:
                p3 += 1
            if dp[i] == dp[p5] * 5:
                p5 += 1
        return dp[n-1]

if __name__ == '__main__':
    sol = Solution()

    assert sol.nthUglyNumber(1) == 1
    assert sol.nthUglyNumber(2) == 2
    assert sol.nthUglyNumber(3) == 3
    assert sol.nthUglyNumber(4) == 4
    assert sol.nthUglyNumber(5) == 5
    assert sol.nthUglyNumber(6) == 6
    assert sol.nthUglyNumber(7) == 8
    assert sol.nthUglyNumber(10) == 12
    assert sol.nthUglyNumber(11) == 15
